🎭 Added text describing `Astro.self` behaviour (#499)
withastro/docs: src/pages/en/reference/api-reference.md

<div>Hello world!</div>
```

### `Astro.self`

> `Astro.self` allows Astro components to be recursively called. This behaviour lets you call the same Astro component over again by defining `<Astro.self>` within the component template.  Allowing you to easily iterate over large data stores and nested data-structures.

Reviewer: ```suggestion
`Astro.self` allows Astro components to be recursively called. This behaviour lets you call the same Astro component over again by defining `<Astro.self>` within the component template. 
```
Do you think maybe these two sentences say pretty much the same thing? Can this be condensed into one sentence?

b) And, I don't think I'd say we've "defined" Astro.self? I know you use "invoke" somewhere else. Depending on the sentence, maybe "called" or "rendered"? Something else?

> `Astro.self` allows Astro components to be recursively called. This behaviour lets you call the same Astro component over again by defining `<Astro.self>` within the component template.  Allowing you to easily iterate over large data stores and nested data-structures.

Reviewer: ```suggestion
within the component template, allowing you to easily iterate over large data stores and nested data-structures.
```
I love the last bit, but it's not a sentence on its own. It's either going to be after a comma fromt whatever you decide for before, or it would need to be something like:

"This allows you to easily iterate..."

```astro
---
const {frontmatter, nestedHeaders} = Astro.props 

> const Headers = frontmatter.headers || nestedHeaders

Reviewer 2: It might be useful for the example to simplify this. Or even better can we just inline the data we're thinking about so that it's easy to see?

So declare some recursive data in the front matter and then use the template Astro.self to render it?

Oh wait. Or we can't do that can we? Because the self needs to receive props. In that case maybe show the component in a separate code block being passed the data?

<div>Hello world!</div>
```

### `Astro.self`

> within the component template, allowing you to easily iterate over large data stores and nested data-structures.

Author: 😵‍💫🧐
I think I made a oops, I had tried to resolve both the editions from Sarah, but only one got applied 😑

> within the component template, allowing you to easily iterate over large data stores and nested data-structures.

Reviewer 2: ```suggestion
`Astro.self` allows Astro components to be recursively called. This behaviour lets you render an Astro component from within itself by using `<Astro.self>` in the component template. This can be helpful for iterating over large data stores and nested data-structures.
```

```astro
---
// NestedHeadings.astro
const { headings } = Astro.props;
---
<ul class="nested-headings">
  <li>{headings.map((item) => {
    if (Array.isArray(item)) {
      // If there is a nested data-structure we render `<Astro.self>`
      // and can pass props through with the recursive call
      return <Astro.self headings={item} />;
    } else {
      return item;
    }
  })}</li>
</ul>
```

This component could then be used like this:

```astro
---
import NestedHeadings from './NestedHeadings.astro';
---
<NestedHeadings headings={['A', ['B', 'C'], 'D']} />
```

And would render HTML like this:

```html
<ul class="nested-headings">
  <li>A</li>
  <li>
    <ul class="nested-headings">
      <li>B</li>
      <li>C</li>
    </ul>
  </li>
  <li>D</li>
</ul>

> ```

Reviewer 2: Sorry for vacillating here @aFuzzyBear. I think maybe having this just be “list” instead of the more specific headings might be clearer?

~~~suggestion
```astro
---
// NestedList.astro
const { items } = Astro.props;
---
<ul class="nested-list">
  <li>{items.map((item) => {
    if (Array.isArray(item)) {
      // If there is a nested data-structure we render `<Astro.self>`
      // and can pass props through with the recursive call
      return <Astro.self items={item} />;
    } else {
      return item;
    }
  })}</li>
</ul>
```

This component could then be used like this:

```astro
---
import NestedList from './NestedList.astro';
---
<NestedList items={['A', ['B', 'C'], 'D']} />
```

And would render HTML like this:

```html
<ul class="nested-list">
  <li>A</li>
  <li>
    <ul class="nested-list">
      <li>B</li>
      <li>C</li>
    </ul>
  </li>
  <li>D</li>
</ul>
```
~~~

Author: no problem in the slightest, list was my first thought, but thoughts of, perhaps a usable example, but given he code demo, list feels like it would suit the generic format you've provided for us dude

> ```

Reviewer 2: Sorry for vacillating here @aFuzzyBear. I think maybe having this just be “list” instead of the more specific headings might be clearer?

~~~suggestion
```astro
---
// NestedList.astro
const { items } = Astro.props;
---
<ul class="nested-list">
  <li>{items.map((item) => {
    if (Array.isArray(item)) {
      // If there is a nested data-structure we render `<Astro.self>`
      // and can pass props through with the recursive call
      return <Astro.self items={item} />;
    } else {
      return item;
    }
  })}</li>
</ul>
```

This component could then be used like this:

```astro
---
import NestedList from './NestedList.astro';
---
<NestedList items={['A', ['B', 'C'], 'D']} />
```

And would render HTML like this:

```html
<ul class="nested-list">
  <li>A</li>
  <li>
    <ul class="nested-list">
      <li>B</li>
      <li>C</li>
    </ul>
  </li>
  <li>D</li>
</ul>
```
~~~
